rfcs/2022-02-23-9481-config-schema.md

```diff
@@ -0,0 +1,823 @@
+supporting types:
+
+```rust
+/// The shape of the field.
+///
+/// This maps similiar to the concept of JSON's data types, where types are generalized and have
+/// generalized representations. This allows us to provide general-but-relevant mappings to core
+/// types, such as integers and strings and so on, while providing escape hatches for customized
+/// types that may be encoded and decoded via "normal" types but otherwise have specific rules or
+/// requirements.
+///
+/// Additionally, the shape of a field can encode some basic properties about the field to which it
+/// is attached. For example, numbers can be bounded on or the lower or upper end, while strings
+/// could define a minimum length, or even an allowed pattern via regular expressions.
+///
+/// In this way, they describe a more complete shape of the field than simply the data type alone.
+#[derive(Clone)]
+pub enum Shape {
+    Null,
+    Boolean,
+    String(StringShape),
+    Number(NumberShape),
+    Array(ArrayShape),
+    Map(MapShape),
+    Composite(Vec<Shape>),
+}
+
+#[derive(Clone, Default)]
+pub struct StringShape {
+    minimum_length: Option<usize>,
+    maximum_length: Option<usize>,
+    allowed_pattern: Option<&'static str>,
+}
+
+#[derive(Clone)]
+pub enum NumberShape {
+    Unsigned {
+        effective_lower_bound: u128,
+        effective_upper_bound: u128,
+    },
+    Signed {
+        effective_lower_bound: i128,
+        effective_upper_bound: i128,
+    },
+    FloatingPoint {
+        effective_lower_bound: f64,
+        effective_upper_bound: f64,
+    }
+}
+
+#[derive(Clone)]
+pub struct ArrayShape {
+    element_shape: Box<Shape>,
+    minimum_length: Option<usize>,
+    maximum_length: Option<usize>,
+}
+
+#[derive(Clone)]
+pub enum MapShape {
+    Fixed(HashMap<&'static str, Shape>),
+    Dynamic(Box<Shape>),
+}
+
+pub struct Field {
+    name: &'static str,
+    description: &'static str,
+    shape: Shape,
+    fields: Vec<Field>,
+    metadata: Option<Vec<Metadata>>,
+}
+
+pub enum TypedMetadata<T: Serialize> {
+    DefaultValue(T),
+    CustomAttribute(String, String),
+}
+
+pub enum Metadata {
+    DefaultValue(Value),
+    CustomAttribute(String, String),
+}
+
+pub trait Configurable<'de>: Serialize + Deserialize<'de> + Sized
+where
+    Self: Clone,
+{
+    /// Gets the human-readable description of this value, if any.
+    ///
+    /// For standard types, this will be `None`. Commonly, custom types would implement this
+    /// directly, while fields using standard types would provide a field-specific description that
+    /// would be used instead of the default descrption.
+    fn description() -> Option<&'static str>;
+
+    /// Gets the shape of this value.
+    fn shape() -> Shape;
+
+    /// Gets the metadata for this value.
+    fn metadata() -> Option<Vec<TypedMetadata<Self>>>;
+
+    /// The fields for this value, if any.
+    fn fields(overrides: Option<Vec<TypedMetadata<Self>>>) -> Option<Vec<Field>>;
```

Comment:
Can there be multiple fields with the same name? If not, a better representation might be a map or set.

Reply:
Thinking this through: no, there could not be multiple instances of the same field name, so yes, this should be a map instead.

Reply:
Addressed this in c9b543cd0.